File: problem19.php
<?php
//solution: 171
echo eulerProblem19() . "\n";

function eulerProblem19() {
    
    $daysInMonths = array(31,28,31,30,31,30,31,31,30,31,30,31);
    $dayOfWeek = "tuesday";
    $sundaysOnFirst = 0;
    for($year = 1901; $year <= 2000; ++$year) {
        for($month = 0; $month < 12; ++$month) {
            
            $dim;
            if($month == 1) {
                $dim = daysInFebruary($year);
            } else {
                $dim = $daysInMonths[$month];
            }
            
            for($day = 1; $day <= $dim; ++$day) {
                if($day == 1 && $dayOfWeek == "sunday") {
                    ++$sundaysOnFirst;
                }
                $dayOfWeek = nextDayOfWeek($dayOfWeek);
            }
        }
    }
    
    return $sundaysOnFirst;
}

function daysInFebruary($year) {
    if($year % 400 == 0) {
        return 29;
    }
    
    if($year % 100 == 0) {
        return 28;
    }
    
    if($year % 4 == 0) {
        return 29;
    }
    
    return 28;
}

function nextDayOfWeek($name) {
    if($name == "monday") {
        return "tuesday";
    } elseif($name == "tuesday") {
        return "wednesday";
    } elseif($name == "wednesday") {
        return "thursday";
    } elseif($name == "thursday") {
        return "friday";
    } elseif($name == "friday") {
        return "saturday";
    } elseif($name == "saturday") {
        return "sunday";
    } elseif($name == "sunday") {
        return "monday";
    }
}

?>

File: python/problem19.py
#!/usr/bin/python
#solution: 171

def euler_problem_19():
    firstmondays = 0
    year = 1900
    dayofweek = "monday"
    daymap = {
        "monday": "tuesday",
        "tuesday": "wednesday",
        "wednesday": "thursday",
        "thursday": "friday",
        "friday": "saturday",
        "saturday": "sunday",
        "sunday": "monday"
    }
    while year < 2001:
        month = [31,28,31,30,31,30,31,31,30,31,30,31]
        for days in month:
            if days == 28 and year % 4 == 0:
                if year % 100 == 0:
                    if year % 400 == 0:
                        days = 29
                else:
                    days = 29
            i = 0
            while i < days:
                if i == 0 and dayofweek == "sunday" and year > 1900:
                    firstmondays += 1
                dayofweek = daymap[dayofweek]
                i += 1
        year += 1

    return firstmondays

print euler_problem_19()
